x-pack/plugins/endpoint/server/routes/resolver/utils/tree.ts

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import { ResolverEvent } from '../../../../common/types';
+import { extractEntityID, extractParentEntityID } from './normalize';
+import { buildPaginationCursor } from './pagination';
+
+interface NodeStats {
+  totalEvents: number;
+  totalAlerts: number;
+}
+
+interface Node {
+  id: string;
+  children: Node[];
+  events: ResolverEvent[];
+  alerts: ResolverEvent[];
+  lifecycle: ResolverEvent[];
+  parent?: Node | null;
+  pagination: {
```

Comment:
I know the comment below talks a little about this but could you add some more comments about when a cursor would be `undefined` vs populated vs `null`?